gradle/documentation/markdown.gradle

```diff
@@ -0,0 +1,190 @@
+ */
+
+import com.vladsch.flexmark.ast.Heading;
+import com.vladsch.flexmark.ext.abbreviation.AbbreviationExtension;
+import com.vladsch.flexmark.ext.attributes.AttributesExtension;
+import com.vladsch.flexmark.ext.autolink.AutolinkExtension;
+import com.vladsch.flexmark.html.HtmlRenderer;
+import com.vladsch.flexmark.parser.Parser;
+import com.vladsch.flexmark.parser.ParserEmulationProfile;
+import com.vladsch.flexmark.util.ast.Document;
+import com.vladsch.flexmark.util.data.MutableDataSet;
+import com.vladsch.flexmark.util.sequence.Escaping;
+
+buildscript {
+  repositories {
+    mavenCentral()
+  }
+
+  dependencies {
+    classpath "com.vladsch.flexmark:flexmark:${scriptDepVersions['flexmark']}"
+    classpath "com.vladsch.flexmark:flexmark-ext-abbreviation:${scriptDepVersions['flexmark']}"
+    classpath "com.vladsch.flexmark:flexmark-ext-attributes:${scriptDepVersions['flexmark']}"
+    classpath "com.vladsch.flexmark:flexmark-ext-autolink:${scriptDepVersions['flexmark']}"
+  }
+}
+
+def getListOfProjectsAsMarkdown = { prefix ->
+  def projects = allprojects.findAll{ it.path.startsWith(prefix) && it.tasks.findByName('renderSiteJavadoc') }
+    .sort(false, Comparator.comparing{ (it.name != 'core') as Boolean }
+      .thenComparing(Comparator.comparing{ (it.name == 'test-framework') as Boolean })
+      .thenComparing(Comparator.comparing{ it.path }));
+  return projects.collect{ project ->
+    def text = "**[${project.path.substring(prefix.length()).replace(':','-')}](${project.relativeDocPath}/index.html):** ${project.description}"
+    if (project.name == 'core') {
+      text = text.concat(' {style="font-size:larger; margin-bottom:.5em"}')
+    }
+    return '* ' + text;
+  }.join('\n')
+}
+
+configure(subprojects.findAll { it.path == ':lucene' || it.path == ':solr' }) {
+  task markdownToHtml(type: Copy) {
+    filteringCharset = 'UTF-8'
+    includeEmptyDirs = false
+    into project.docroot
+    rename(/\.md$/, '.html')
+    filter(MarkdownFilter)
+  }
+}
+
+configure(project(':lucene')) {
+  markdownToHtml {
+    from('.') {
+      include 'MIGRATE.md'
+      include 'JRE_VERSION_MIGRATION.md'
+      include 'SYSTEM_REQUIREMENTS.md'
+    }
+  }
+  
+  task createDocumentationIndex {
+    def outputFile = file("${project.docroot}/index.html");
+    def defaultCodecFile = project(':lucene:core').file('src/java/org/apache/lucene/codecs/Codec.java')
+    
+    inputs.file(defaultCodecFile)
+    outputs.file(outputFile)
+    
+    doLast {
+      // static Codec defaultCodec   =   LOADER    .   lookup    (   "LuceneXXX"  )   ;
+      def regex = ~/\bdefaultCodec\s*=\s*LOADER\s*\.\s*lookup\s*\(\s*"([^"]+)"\s*\)\s*;/
+      def matcher = regex.matcher(defaultCodecFile.getText('UTF-8'))
+      if (!matcher.find()) {
+        throw GradleException("Cannot determine default codec from file ${defaultCodecFile}")
+      }
+      def defaultCodecPackage = matcher.group(1).toLowerCase(Locale.ROOT)
+      def markdown = """
```

Comment:
Should we move it to an external file? Then we can have a proper suffix and editor support.

Reply:
That was my question. Can we include it in a way so the groovy variables are parsed?

Reply:
https://docs.groovy-lang.org/latest/html/api/groovy/text/SimpleTemplateEngine.html
Should work?

Reply:
Cool. So I will place the file next to the legacy XSL file used by Ant. Once Ant is retired, the latter can go away.